oci/archive/reader.go

```diff
@@ -0,0 +1,93 @@
+package archive
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/image/v5/internal/tmpdir"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type reader struct {
+	manifest *imgspecv1.Index
+	tempDirOCIRef
+}
+
+func CreateUntarTempDirReader(src string, sys *types.SystemContext) (*reader, error) {
+	// TODO: This can take quite some time, and should ideally be cancellable using a context.Context.
+	arch, err := os.Open(src)
+	if err != nil {
+		return nil, err
+	}
+	defer arch.Close()
+
+	dst, err := ioutil.TempDir(tmpdir.TemporaryDirectoryForBigFiles(sys), "oci")
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp directory")
+	}
+
+	reader := reader{
+		tempDirOCIRef: tempDirOCIRef{tempDirectory: dst},
+	}
+
+	if err := archive.NewDefaultArchiver().Untar(arch, dst, &archive.TarOptions{NoLchown: true}); err != nil {
+		if err := reader.tempDirOCIRef.deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory %q", dst)
+		}
+		return nil, errors.Wrapf(err, "error untarring file %q", dst)
+	}
+
+	indexJSON, err := os.Open(filepath.Join(dst, "index.json"))
+	if err != nil {
+		return nil, err
+	}
+	defer indexJSON.Close()
+	reader.manifest = &imgspecv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(reader.manifest); err != nil {
+		return nil, err
+	}
+
+	return &reader, nil
+}
+
+// ReferenceWrapper includes the reference for the image and image dest name
+type ReferenceWrapper struct {
+	DestName       string
+	ImageReference types.ImageReference
+}
+
+// List returns a list of ReferenceWrapper for images in the reader.
+// the ImageReferences of the ReferenceWrapper are valid only until the Reader is closed.
+func (r *reader) List() ([]ReferenceWrapper, error) {
+	res := []ReferenceWrapper{}
+	for _, md := range r.manifest.Manifests {
+		if md.MediaType != imgspecv1.MediaTypeImageManifest && md.MediaType != imgspecv1.MediaTypeImageIndex {
+			continue
+		}
+		refName, ok := md.Annotations[imgspecv1.AnnotationRefName]
```

Comment:
I think we need to support more than one name as we support with docker-archive.

Doing a `podman save -o foo.tar image1-a image1-b image1-c image2` with image1-* all referring to the same image, the names will be preserved.

Reply:
The names of the same image are preserved as `imgspecv1.AnnotationRefName`("org.opencontainers.image.ref.name") field of different entries. And I changed the `ReferenceWrapper` https://github.com/containers/image/pull/1072#discussion_r516612025, use a map to keep the name and ImageReferences.